src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/AddPackageCommandUtility.cs

```diff
@@ -88,9 +88,10 @@ public static async Task<NuGetVersion> GetLatestVersionFromSourceAsync(PackageSo
         /// Returns the PackageSource with its credentials if available
         /// </summary>
         /// <param name="requestedSources">Sources to match</param>
+        /// <param name="additionalSources">Additional user supplied source feeds as argument</param>
```

Comment:
I don't think they are supposed to be additional, but rather they are supposed to replace the existing sources.

Reply:
I believe it's treated as addition into existing sources user already setup on project. 
Here below screenshot case it already has nuget.org as source and also `Mysource` which is I passed as argument on [AddPackageReferenceCommandRunner.cs#L347](https://github.com/NuGet/NuGet.Client/blob/f9ad775ce70967f23338566b806b7bde1f623ec3/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs#L347). Maybe user can use this one as alternative sourcefeed for just once during restore.

![image](https://user-images.githubusercontent.com/8766776/101226533-9d082000-3649-11eb-9d53-9cb86c014cad.png)

Reply:
I don't think this is a technical question. 

`--source` for restore *completely* overwrites the sources. 

`dotnet add package` should work the same. 

There are docs that describe this behavior as well: https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-add-package .

Reply:
@nkolev92 
Can you suggest name for it? 
How about `Overriding source` or `Replacing source`, `New sources`?

Reply:
The sources in the restore request are not used at all. 
The sources in the dg spec are used https://github.com/NuGet/NuGet.Client/blob/6562ef31f891c26981f5f5cc82f3fe5c21451eb9/src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreArgs.cs#L129-L144. 

Don't think you'd need to change it there. It'd be difficult to do without consequences, so I'd just change in the package spec that you are passing for restore.

Reply:
Let's sync on Thursday if you are still stuck.

Reply:
@nkolev92 
Yes, can we talk 5-10 min today. Early is better.

Reply:
@nkolev92 
I tested example on [https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-add-package](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-add-package):
`dotnet add package Microsoft.AspNetCore.StaticFiles -s https://dotnet.myget.org/F/dotnet-core/api/v3/index.json
`

Reply:
It looks one problem with current behavior of `--source` we can't do repeatable restore/build if local cache is cleared even though we successfully added package from custom relative source unless that one is added into source list.

Reply:
> It looks one problem with current behavior of --source we can't do repeatable restore

That's exactly why source is not supported in the PM UI & why you can't select a subset of sources when doing restore, it always uses them all. 

I think it's *ok* if you overwrite the source in this case. 
If the customer gets themselves in a non-repeatable position, it is what it is. 
They can already do that with /p:RestoreSources when they pass something different from what's specified in their nuget.config. 

> It doesn't change the source property on dgspec file. It's only used as temporary thing. It's not persisted, I can't say --source completely overwrites the sources.

I'm confused what this means. It doesn't work at all, it's disregarding it, which is what you are trying to fix.

Reply:
> I think it's _ok_ if you overwrite the source in this case.

@nkolev92 

I think overwriting is not good idea, maybe append is better. 

First of all let's run this:
`dotnet restore /p:restoreSources="C:\NuGetProj\IssueRepro\5127\Mypackages3;C:\NuGetProj\IssueRepro\5127\Mysource2"`

```
   Csproj file content: 

  <ItemGroup>
    <PackageReference Include="Domain3" Version="1.0.0" />
    <PackageReference Include="Domain2" Version="1.0.0" />
  </ItemGroup>
```

#### RestoreSuccessful
and source changes to below:

```
        "sources": {
          "C:\\NuGetProj\\IssueRepro\\5127\\Mypackages3": {},
          "C:\\NuGetProj\\IssueRepro\\5127\\Mysource2": {}
        },

```

### Another question is do we write relative path or absolute path in case  user passed relative path for source?

```
        "sources": {
          "..\Mysource2": {}
        },

```
or 

```
        "sources": {
          "C:\\NuGetProj\\IssueRepro\\5127\\Mysource2": {}
        },

```

### Overwrite

By default usually we have this source:
        "sources": {
          "https://api.nuget.org/v3/index.json": {}
        },

`dotnet add package Domain2 -s ..\Mysources2`

After overwrite we going to have 
```
        "sources": {
          "..\Mysources2": {},
        },
```

`dotnet add package Domain3 -s ..\Mypackages3`

After yet another overwrite we going to have 
```
        "sources": {
          "..\\Mypackages3": {},
        },
```

As you can see it's not great since every action overwrites the previous source.

### Append.

By default usually we have this source:
        "sources": {
          "https://api.nuget.org/v3/index.json": {}
        },

`dotnet add package Domain2 -s ..\Mysources2`

After append we going to have 
```
        "sources": {
          "..\Mysources2": {},
          "https://api.nuget.org/v3/index.json": {}
        },
```

`dotnet add package Domain3 -s ..\Mypackages3`

After append we going to have 
```
        "sources": {
          "..\Mysources2": {},
          "..\\Mypackages3": {},
          "https://api.nuget.org/v3/index.json": {}
        },
```

This looks better, because it looks similar to [RestoreSuccessful](#RestoreSuccessful).
Only issue is we have `"https://api.nuget.org/v3/index.json": {}` which is not in [RestoreSuccessful](#RestoreSuccessful).

Please give your thought?